river/cluster/evostream.py

```diff
@@ -0,0 +1,424 @@
+            else:
+                merged_mcs[count] = micro_clusters[i]
+                for j in range(i + 1, n_micro_cluster):
+                    if (not merged_status[j]) and self._distance(
+                        merged_mcs[count].center, micro_clusters[j].center
+                    ) <= self.radius:
+                        merged_mcs[count].merge(micro_clusters[j])
+                        merged_status[j] = True
+                count += 1
+        return merged_mcs
+
+    def _cleanup(self):
+        # Algorithm 2 of Michael Hahsler and Matthew Bolanos: Cleanup process to remove
+        # inactive clusters and shared density entries from memory
+
+        for i in list(self._micro_clusters):
+            self._micro_clusters[i].weight = self._micro_clusters[i].weight * (
+                2
+                ** (
+                    -self.decay_rate
+                    * (self.time_stamp - self._micro_clusters[i].last_update)
+                )
+            )
+            if self._micro_clusters[i].weight < 2 ** (
+                -self.decay_rate * self.cleanup_interval
+            ):
+                self._micro_clusters.pop(i)
+
+        self._micro_clusters = {
+            i: v for i, v in enumerate(self._micro_clusters.values())
+        }
+
+        self._micro_clusters = self._merge_micro_clusters(self._micro_clusters)
+
+    @staticmethod
+    def _generate_population(micro_clusters, population_size, n_clusters):
+        cluster_population = {}
+        for i in range(population_size):
+            cluster_population_i_keys = random.sample(micro_clusters.keys(), n_clusters)
+            cluster_population[i] = {
+                i: micro_clusters[cluster_population_i_keys[i]]
+                for i in range(n_clusters)
+            }
+        return cluster_population
+
+    def learn_one(self, x, sample_weight=None):
+
+        self._update(x)
+
+        if self.time_stamp % self.cleanup_interval == 0:
+            self._cleanup()
+
+        # initialize macro clusters
+        if (
+            len(self._micro_clusters) == self.initialization_threshold
+            and not self.initialized
+        ):
+            self._cluster_population = self._generate_population(
+                self._micro_clusters, self.initialization_threshold, self.n_clusters
+            )
+            self.initialized = True
+
+        if self.time_stamp % self.evolution_interval == 0:
+            self._evolution()
+
+            clusters_solution_index = np.argmax(
+                list(self._fitness_cluster_population.values())
+            )
+
+            self.clusters = self._cluster_population[clusters_solution_index]
+
+            self.centers = {i: self.clusters[i].center for i in self.clusters}
+
+        return self
+
+    def predict_one(self, x, sample_weight=None):
+
+        label = self._find_closest_cluster_index(x, self.clusters)
+
+        return label
+
+
+class evoStreamMicroCluster(metaclass=ABCMeta):
```

Comment:
Why this class is linked to `ABCMeta`?

Besides that, as a possible performance trick, you check the usage of `__slots__` in this class :)

Reply:
Usually, for microcluster structures of all online clustering algorithms, we always link this to `ABCMeta`. Maybe @jacobmontiel can you provide a better explanation on this?